### PR TITLE
Use paths relative to the parameter file for other material files

### PR DESCRIPTION
**Pull request by davschneller**
Cf. the suggestions in #1154 .
(still needs testing, hence draft)

This PR changes the logic of finding paths as parameters:

* first, we look for the file relative to the parameter file. If it is found, we use that file.
* otherwise, we default to the current execution directory (as it was before). Currently, we don't fail if that file path does not exist (but we could, in principle—the only stumbling stone would be the Netcdf file paths normally not having the `.nc` file ending in the parameter file).

Also, we now work with the absolute file paths after having read them; i.e. they'll also show up in the log like that.

### CODE REVIEW

File: src/Initializer/Parameters/ParameterReader.cpp [hunk 1]
```diff
@@ -1,7 +1,51 @@
 #include "ParameterReader.h"
 
+#include "Common/filesystem.h"
```
[the remark this answers is not in the document]
Comment (davschneller) [reply src/Initializer/Parameters/ParameterReader.cpp:8]: It does, yes—but it respects empty lines in between right now :)
(in principle, that section here can be merged though)

File: src/Initializer/Parameters/ParameterReader.h [hunk 2]
```diff
@@ -24,18 +24,27 @@ inline void sanitize(std::string& input) {
...
 
   template <typename T>
-  T readWithDefault(const std::string& field, const T& defaultValue) {
-    T value = defaultValue;
+  std::optional<T> read(const std::string& field) {
     if (hasField(field)) {
-      value = readUnsafe<T>(field);
+      return readUnsafe<T>(field);
+    } else {
+      return std::optional<T>();
```
[the remark this answers is not in the document]
Comment (davschneller) [reply src/Initializer/Parameters/ParameterReader.h:34]: Sometimes it's desirable to have an empty optional instead—see e.g. the `readWithDefault` below. In effect the `readWithDefault(param, default)` call is now equivalent to `read(param).or_else(default)` .

File: src/Initializer/Parameters/ParameterReader.cpp [hunk 3]
```diff
@@ -1,11 +1,56 @@
...
+        return nextPath;
+      }
+    }();
+    // try to get the full file name (if that works)
+    if (filesystem::exists(loadFileName)) {
+      return filesystem::canonical(loadFileName);
+    } else if (filesystem::exists(nextPath)) {
+      return filesystem::canonical(nextPath);
+    } else {
+      // otherwise, just return the string (TODO: or should be fail here then?)
```
[the remark this answers is not in the document]
Comment (davschneller) [reply src/Initializer/Parameters/ParameterReader.cpp:34]: We can do that here.

The only thing that currently still complicates that would be that the Netcdf files so far did not need a file ending.
However, iirc they are only ever used in case of convergence tests at this point in time (because they support periodic boundary conditions, and PUML doesn't—yet), so we could just enforce the file endings now. Or alternatively, supply an optional file ending to this function.

File: src/Initializer/Parameters/ParameterReader.cpp [hunk 4]
```diff
@@ -1,11 +1,56 @@
...
+      return nextPath;
+    }
+  } else {
+    return std::optional<std::string>();
```
[the remark this answers is not in the document]
Comment (davschneller) [reply src/Initializer/Parameters/ParameterReader.cpp:38]: (cf. the comment below)

**Review by vikaskurapati — APPROVED** [review]
